csrc/multidevice/utils.cpp

```diff
@@ -380,49 +434,36 @@ bool haveDifferentShardings(
       mapDeviceParallelTypeToId(consumer->getLoopDomain());
 
   for (const auto parallel_type : kParallelTypeDIDs) {
-    IterDomain* p_loop_id = getOrDefault(p_parallel_type_to_id, parallel_type);
-    if (p_loop_id != nullptr) {
-      auto p_inputs =
-          getInputsInTargetDomain(p_loop_id, producer->getLogicalDomain());
-      if (!overlaps(p_inputs, mapped_p_logical_ids)) {
-        p_loop_id = nullptr;
-      }
+    IterDomain* p_id = getOrDefault(p_parallel_type_to_id, parallel_type);
+    auto [p_index, p_mapped] =
+        computeIndex(p_id, producer->getLogicalDomain(), known_indices);
+    if (!p_mapped) {
+      p_index = nullptr;
     }
 
-    IterDomain* c_loop_id = getOrDefault(c_parallel_type_to_id, parallel_type);
-    if (c_loop_id != nullptr) {
-      auto c_inputs =
-          getInputsInTargetDomain(c_loop_id, consumer->getMaybeRootDomain());
-      if (!overlaps(c_inputs, mapped_c_root_ids)) {
-        c_loop_id = nullptr;
-      }
+    IterDomain* c_id = getOrDefault(c_parallel_type_to_id, parallel_type);
+    auto [c_index, c_mapped] =
+        computeIndex(c_id, consumer->getMaybeRootDomain(), known_indices);
+    if (!c_mapped) {
+      c_index = nullptr;
     }
 
-    auto is_mapped_in_id_model =
-        [](IterDomain* a, IterDomain* b, const IdModel& id_model) -> bool {
-      if (a == nullptr && b == nullptr) {
+    const bool is_equivalent = [&]() -> bool {
+      if (p_index == nullptr && c_index == nullptr) {
         return true;
       }
-
-      if (a == nullptr || b == nullptr) {
+      if (p_index == nullptr || c_index == nullptr) {
         return false;
       }
 
-      // Going between bDIDx{1} and iDIDx{N} doesn't trigger resharding, but
-      // would be flagged by ALMOSTEXACT as a false positive.
-      if (id_model.idGraph(IdMappingMode::BROADCAST)
-              .disjointValSets()
-              .strictAreMapped(a, b)) {
-        return true;
-      }
-
-      // Check ALMOSTEXACT so iDIDx{N}*b{1} and iDIDx{N} are mapped.
-      return id_model.idGraph(IdMappingMode::ALMOSTEXACT)
-          .disjointValSets()
-          .strictAreMapped(a, b);
-    };
+      Val* const p_index_simplified = simplifyExpr(p_index, {}, assumptions);
+      Val* const c_index_simplified = simplifyExpr(c_index, {}, assumptions);
+      // I chose to not inline `p_index_simplified` and `c_index_simplified` so
+      // it's easier to print them during debugging.
+      return p_index_simplified->sameAs(c_index_simplified);
```

Comment:
Would it be more robust to do the following?
```C++
Val* is_same = simplifyExpr(SimplifyingIrBuilder::eqExpr(p_index, c_index), {}, assumptions);
return is_same->isTrue();
```
This is better because, we may have:
```C++
p_index = x + y;
c_index = y + x;
```
They are both simplified most, but, but they are not `sameAs` of each other.